src/print/index.ts

```diff
@@ -530,174 +526,152 @@ function printAttributeNodeValue(
-        const childNode = childPath.getValue() as Node;
-        const childDoc = childPath.call(print);
-
-        if (isInlineNode(childNode)) {
-            currentGroup.push({ doc: childDoc, node: childNode });
+    if (elementType === 'blockEl') {
+        let line: Doc = softline;
+        if (firstChild === lastChild && firstChild.type === 'Text') {
+            trimTextNodeLeft(firstChild);
+            trimTextNodeRight(firstChild);
         } else {
-            flush();
-
-            if (childDoc !== '') {
-                outputChildDoc(
-                    isLine(childDoc) ? childDoc : concat([breakParent, childDoc]),
-                    childNode,
-                );
+            if (isTextNodeStartingWithLinebreak(firstChild)) {
+                trimTextNodeLeft(firstChild);
+                line = hardline;
+            }
+            if (isTextNodeEndingWithLinebreak(lastChild)) {
+                trimTextNodeRight(lastChild);
             }
         }
-    }, 'children');
-
-    flush();
-    lastChildDocProduced();
-
-    return childDocs;
-}
-
-/**
- * Print the nodes in `path` indented and with leading and trailing newlines.
- */
-function printIndentedWithNewlines(path: FastPath, print: PrintFn): Doc {
-    return indent(
-        concat([softline, ...trim(printChildren(path, print), isLine), dedent(softline)]),
-    );
-}
 
-/**
- * Print the nodes in `path` indented but without adding any leading or trailing newlines.
- */
-function printIndentedPreservingWhitespace(path: FastPath, print: PrintFn) {
-    return indent(concat(dedentFinalNewline(printChildren(path, print))));
+        return concat([
+            indent(
+                concat([
+                    line,
+                    ...path.map(print, 'children'),
+                    // TODO not that simple unfortunately: only break around block elements when there's more than one
+                    // hasOnlyTextAndMoustacheChildren ? '' : breakParent,
+                ]),
+            ),
+            line,
+        ]);
+    } else {
+        if (
+            firstChild !== lastChild &&
+            isTextNodeStartingWithLinebreak(firstChild) &&
+            isTextNodeEndingWithLinebreak(lastChild)
+        ) {
+            trimTextNodeLeft(firstChild);
+            trimTextNodeRight(lastChild);
+            return concat([indent(concat([hardline, ...path.map(print, 'children')])), hardline]);
+        }
+        return concat(path.map(print, 'children'));
+    }
 }
 
 /**
  * Split the text into words separated by whitespace. Replace the whitespaces by lines,
  * collapsing multiple whitespaces into a single line.
  *
- * If the text starts or ends with multiple newlines, those newlines should be "keepIfLonely"
- * since we want double newlines in the output.
+ * If the text starts or ends with multiple newlines, two of those should be kept.
  */
-function splitTextToDocs(text: string): Doc[] {
+function splitTextToDocs(node: TextNode): Doc[] {
```

Comment:
This function could benefit from comments. I don't understand the different cases.